algorithms/nautilus/grammartec/tree.cpp

```diff
@@ -0,0 +1,618 @@
+ * @param (n) Node ID
+ * @param (ctx) Context
+ * @throw std::out_of_range Node ID is invalid
+ * @return Rule corresponding to node ID
+ */
+const Rule& Tree::GetRule(const NodeID& n, Context& ctx) const {
+  return ctx.GetRule(GetRuleID(n));
+}
+
+/**
+ * @fn
+ * Get the custom rule data of a node.
+ * @brief Get custom rule data by node ID
+ * @param (n) Node ID
+ * @throw std::out_of_range Node ID is invalid
+ * @return Data of rule (throws exception if rule is not Custom)
+ */
+const std::string& Tree::GetCustomRuleData(const NodeID& n) const {
+  return _rules.at(static_cast<size_t>(n)).Data();
+}
+
+/**
+ * @fn
+ * Get the RuleIROrCustom instance of a node.
+ * @brief Get rule ID or custom by node ID
+ * @param (n) Node ID
+ * @throw std::out_of_range Node ID is invalid
+ * @return RuleIDOrCustom corresponding to node ID
+ */
+const RuleIDOrCustom& Tree::GetRuleOrCustom(const NodeID& n) const {
+  return _rules.at(static_cast<size_t>(n));
+}
+
+/**
+ * @fn
+ * Clear every rule, size, and parent information.
+ * @brief Remove every rule
+ */
+void Tree::Truncate() {
+  _rules.clear();
+  _sizes.clear();
+  _paren.clear();
+}
+
+/**
+ * @fn
+ * Generate a random tree from a nonterminal symbol.
+ * @brief Generate tree from nonterminal
+ * @param (start) Nonterminal symbol ID
+ * @param (len) Maximum length for getting random rule for `start`
+ * @param (ctx) Context
+ */
+void Tree::GenerateFromNT(const NTermID& start, size_t len, Context& ctx) {
+  const RuleID& rid = ctx.GetRandomRuleForNT(start, len);
+  GenerateFromRule(rid, len - 1, ctx);
+}
+
+/**
+ * @fn
+ * Generate a random tree from a specific rule of a nonterminal.
+ * @brief Generate tree from rule
+ * @param (ruleid) Rule ID
+ * @param (max_len) Maximum length
+ * @param (ctx) Context
+ */
+void Tree::GenerateFromRule(const RuleID& ruleid, size_t max_len, Context& ctx) {
+  if (std::holds_alternative<PlainRule>(ctx.GetRule(ruleid).value())) {
+
+    /* PlainRule or ScriptRule */
+    Truncate();
+    _rules.emplace_back(ruleid);
+    _sizes.emplace_back(0);
+    _paren.emplace_back(0);
+    ctx.GetRule(ruleid).Generate(*this, ctx, max_len);
+
+    _sizes[0] = _rules.size();
+
+  } else {
+
+    // NOTE: RegExpRule should work differently here
+    throw exceptions::not_implemented(
+      "Only PlainRule is supported", __FILE__, __LINE__
+    );
+
+  }
+}
+
+/**
+ * @fn
+ * Calculate the recursions of this tree.
+ * @brief Calculate recursions
+ * @param (ctx) Context
+ * @return Vector of recursion info if successful (std::optional)
+ */
+std::optional<std::vector<RecursionInfo>> Tree::CalcRecursions(Context& ctx) {
+  std::vector<RecursionInfo> ret;
+  std::unordered_set<NTermID> done_nterms;
+
+  for (const RuleIDOrCustom& rule: _rules) {
+    const NTermID& nterm = ctx.GetNT(rule);
+    if (done_nterms.find(nterm) == done_nterms.end()) {
+      try {
+        ret.emplace_back(*this, nterm, ctx);
+      } catch (const char*) {
+        // pass
+      }
```

Comment:
What's this try-catch block?

Reply:
A-ha now I see.
As much as possible, please try not to use exception. Exceptions are extremely slow in general.
You can create some function like Rust's `::new` method instead.

Reply:
Fixed in [da9dbdb](https://github.com/fuzzuf/fuzzuf/pull/23/commits/da9dbdbabb27b0647b085c9a8d471a7328ee2e37)

Reply:
Re-designed in [da7fd83](https://github.com/fuzzuf/fuzzuf/pull/23/commits/da7fd8366807b6e8bb8a55a425b8c8b0000babae)

Reply:
Excellent.